components/brave_wallet/browser/brave_wallet_pin_service.cc

```diff
@@ -0,0 +1,778 @@
+    return nullptr;
+  }
+  if (parts.at(1) == kLocalService) {
+    return absl::nullopt;
+  } else {
+    return parts.at(1);
+  }
+}
+
+void BraveWalletPinService::Validate(BlockchainTokenPtr token,
+                                     const absl::optional<std::string>& service,
+                                     ValidateCallback callback) {
+  mojom::TokenPinStatusPtr status = GetTokenStatus(service, token);
+  if (!status) {
+    std::move(callback).Run(false, nullptr);
+    return;
+  }
+  if (status->code != mojom::TokenPinStatusCode::STATUS_PINNED) {
+    std::move(callback).Run(false, nullptr);
+    return;
+  }
+
+  absl::optional<std::vector<std::string>> cids =
+      ResolvePinItems(service, token);
+
+  if (!cids) {
+    SetTokenStatus(service, std::move(token),
+                   mojom::TokenPinStatusCode::STATUS_PINNING_IN_PROGRESS,
+                   nullptr);
+    std::move(callback).Run(true, nullptr);
+    return;
+  }
+
+  auto path = GetPath(absl::nullopt, token);
+  if (!path) {
+    std::move(callback).Run(
+        false,
+        mojom::PinError::New(mojom::WalletPinServiceErrorCode::ERR_WRONG_TOKEN,
+                             "Wrong token data"));
+    return;
+  }
+
+  if (!service) {
+    local_pin_service_->ValidatePins(
+        path.value(), cids.value(),
+        base::BindOnce(&BraveWalletPinService::OnTokenValidated,
+                       base::Unretained(this), service, std::move(callback),
+                       std::move(token)));
+  } else {
+    // Remote pinning not implemented yet
+    std::move(callback).Run(false, nullptr);
+  }
+}
+
+void BraveWalletPinService::IsLocalNodeRunning(
+    IsLocalNodeRunningCallback callback) {
+  std::move(callback).Run(ipfs_service_->IsDaemonLaunched());
+}
+
+void BraveWalletPinService::OnIpfsLaunched(bool result, int64_t pid) {
+  for (const auto& observer : observers_) {
+    observer->OnLocalNodeStatusChanged(result);
+  }
+}
+
+void BraveWalletPinService::OnIpfsShutdown() {
+  for (const auto& observer : observers_) {
+    observer->OnLocalNodeStatusChanged(false);
+  }
+}
+
+void BraveWalletPinService::MarkAsPendingForPinning(
+    const mojom::BlockchainTokenPtr& token,
+    const absl::optional<std::string>& service) {
+  SetTokenStatus(service, token,
+                 mojom::TokenPinStatusCode::STATUS_PINNING_PENDING, nullptr);
+}
+
+void BraveWalletPinService::MarkAsPendingForUnpinning(
+    const mojom::BlockchainTokenPtr& token,
+    const absl::optional<std::string>& service) {
+  SetTokenStatus(service, token,
+                 mojom::TokenPinStatusCode::STATUS_UNPINNING_PENDING, nullptr);
+}
+
+void BraveWalletPinService::AddPin(BlockchainTokenPtr token,
+                                   const absl::optional<std::string>& service,
+                                   AddPinCallback callback) {
+  if (!token->is_nft) {
+    auto pin_error = mojom::PinError::New(
+        mojom::WalletPinServiceErrorCode::ERR_WRONG_TOKEN, "Token is not nft");
```

Comment:
Would this be used in UI in the future? If so, error messages should be localized.

Reply:
No this is not used, just error description for service clients